### PR TITLE
refactor(api): create explicitly public Deck API

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR bites off a chunk of RCORE-420 by sequestering the old, mutable `Deck` class in `opentrons.protocol_api.core.protocol_api` (aka "the legacy core") and creating a new, explicitly public `AbstractDeck` and `Deck` interfaces to be returned from `ProtocolContext.deck`.

Legacy Deck objects will continue to be returned if the legacy PAPI core is in place, so this shouldn't break any existing workflows.

## Changelog

- `ProtocolContext.deck` now returns an instance of `opentrons.protocol_api.Deck` if using the engine-based core
    - If using the legacy core, no change in behavior
- The type signature of `ProtocolContext.deck` has now changed to:
    - `Deck[slot] -> Labware | ModuleContext`
    - This is _not_ what the current Protocol API does
    - The "publicness" of this API is and has been super fuzzy, and RCORE-420 will take a few PRs to complete, so we'll have this typing discrepancy in a better place before we close out the issue

### Short history of `Deck`

The type signature of `Deck` has changed over time, sometimes without any bumps to API version

1. Initially: `Dict[str | int, Labware | None]`
2. Modules added: `Dict[str | int, Labware | ModuleGeometry | None]`
    - It's unclear to me how public `ModuleGeometry` was intended to be at the time
3. Protocol API refactored: `Dict[str | int, LabwareCore | ModuleGeometry | None]`
    - `deck[12]` continued to be a `Labware` in Python protocols
4. This PR: `Dict[str | int, Labware | ModuleContext | None]`
    - Currently guarded by FF at runtime
    - We intend to guard this behavior behind an API version bump
    - `ModuleGeometry` is not really a public API
    - Returning a `Labware` or `ModuleContext` feels more in line with what an API user might expect, based on documentation

## Review requests

I've run protocol smoke tests on OT-2 with the FF on, we should do the same on an OT-3. This PR in isolation may temporarily break some previously working stuff until future PRs in this ticket land.

Old calibration flows on the OT-2 (deck cal, deck cal check, tip length, pipette offset) need to be tested, too.

## Risk assessment

Low. Greatest risk is probably the accidental introduction of a circular dependency somewhere unexpected.

### CODE REVIEW

File: api/src/opentrons/protocol_api/deck.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    id: str
+    position: Tuple[float, float, float]
+    displayName: str
+
+
+class AbstractDeck(ABC, Mapping[DeckLocation, DeckItem]):
+    """Abstract interface for :py:class:`Deck`."""
+
+    @abstractmethod
+    def __getitem__(self, key: DeckLocation) -> DeckItem:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/deck.py]: Probably try to preserve `deck["3"] is None` if slot is empty. A bit funny for a "dict-like" interface, but breaking the existing API here seems unnecessary

```suggestion
class AbstractDeck(ABC, Mapping[DeckLocation, Optional[DeckItem]]):
    """Abstract interface for :py:class:`Deck`."""

    @abstractmethod
    def __getitem__(self, key: DeckLocation) -> Optional[DeckItem]:
```

File: api/src/opentrons/protocol_api/deck.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+class Deck(AbstractDeck):
+    """A dictionary-like object to access Protocol API objects loaded on the deck.
+
+    Accessible via :py:meth:`ProtocolContext.deck`.
+    """
+
+    def __init__(self, protocol_core: ProtocolCore) -> None:
+        self._protocol_core = protocol_core
+
+    def __getitem__(self, key: DeckLocation) -> DeckItem:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/deck.py]:
```suggestion
    def __getitem__(self, key: DeckLocation) -> Optional[DeckItem]:
```

File: api/src/opentrons/protocol_api/core/protocol_api/deck.py [hunk 3]
```diff
@@ -1,47 +1,53 @@
...
+
+    @property
+    def load_name(self) -> str:
+        ...
 
-    id: str
-    position: List[float]
-    displayName: str
 
+class Deck(AbstractDeck, UserDict):  # type: ignore[type-arg]
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/core/protocol_api/deck.py]:
```suggestion
class Deck(AbstractDeck, UserDict[Any, Any]):
```

File: api/src/opentrons/protocol_api/core/protocol_api/deck.py [hunk 4]
```diff
@@ -146,14 +152,14 @@ def is_edge_move_unsafe(self, mount: types.Mount, target: "Labware") -> bool:
...
-        if mount is types.Mount.RIGHT:
+        if mount is Mount.RIGHT:
             other_labware = self.left_of(slot)
         else:
             other_labware = self.right_of(slot)
 
         return isinstance(other_labware, ModuleGeometry)
 
-    def right_of(self, slot: str) -> Optional[DeckItem]:
+    def right_of(self, slot: DeckLocation) -> Optional[DeckItem]:  # type: ignore[override]
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/core/protocol_api/deck.py]: Hm, what's going on with the `# type: ignore[override]:` here?

If I'm reading things correctly:

`core.protocol_api.deck.Deck` is returning a `DeckItem`, where "a `DeckItem`" here is anything with a `highest_z()` and `load_name()` method.

But the superclass, `AbstractDeck`, defines this method as returning a *different* `DeckItem`: `Union[Labware, ModuleContext]`.

Why does this module need to define its own `DeckItem` `Protocol`? Why can't it use `Union[Labware, ModuleContext]` so the types match?

File: api/src/opentrons/protocol_api/core/protocol_api/deck.py [hunk 5]
```diff
@@ -89,18 +95,18 @@ def _check_name(self, key: object) -> int:
...
+            _log.exception("Bad slot name: {}".format(key))
             should_raise = True
         should_raise = should_raise or key_int not in self.data
         if should_raise:
             raise ValueError("Unknown slot: {}".format(key))
         else:
             return key_int
 
-    def __getitem__(self, key: types.DeckLocation) -> DeckItem:
+    def __getitem__(self, key: DeckLocation) -> Optional[DeckItem]:  # type: ignore[override]
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/core/protocol_api/deck.py]: Why is it allowed for us to `# type: ignore` the fact that this method is returning `Optional[DeckItem]` where the `AbstractDeck` interface specifies a non-optional `DeckItem`?

If this is an unavoidable reality of this implementation, then should the `AbstractDeck` interface be modified to match?

File: api/src/opentrons/protocol_api/core/protocol_api/deck.py [hunk 6]
```diff
@@ -265,12 +271,12 @@ def get_calibration_position(self, id: str) -> CalibrationPosition:
             )
         return calibration_position
 
-    def get_fixed_trash(self) -> Optional[Labware]:
+    def get_fixed_trash(self) -> Optional[Union[Labware, LabwareImplementation]]:
         fixtures = self._definition["locations"]["fixtures"]
         ft = next((f for f in fixtures if f["id"] == FIXED_TRASH_ID), None)
-        return self.data[self._check_name(ft.get("slot"))] if ft else None
+        return self.data[self._check_name(ft.get("slot"))] if ft else None  # type: ignore[return-value]
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/core/protocol_api/deck.py]: What changed to necessitate this change in the return type and the new `# type: ignore`? Was this buggy all along and we just have stricter type-checking now?

File: api/src/opentrons/protocol_api/deck.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        position: The absolute x, y, z coordinate of the point.
+        displayName: A human-readable nickname for this point.
+    """
+
+    id: str
+    position: Tuple[float, float, float]
+    displayName: str
+
+
+class AbstractDeck(ABC, Mapping[DeckLocation, Optional[DeckItem]]):
```
Comment (mcous) [inline api/src/opentrons/protocol_api/deck.py]: I'm...  not sure this is totally necessary. I'm going to remove it from this PR and add it back in if it's actually needed

File: api/src/opentrons/protocol_api/core/protocol_api/deck.py [hunk 8]
```diff
@@ -172,16 +178,16 @@ def left_of(self, slot: str) -> Optional[DeckItem]:
...
-        return types.Location(self._positions[key_int], str(key))
+        return Location(self._positions[key_int], str(key))
 
     def recalculate_high_z(self) -> None:
         self._highest_z = 0.0
         for item in [lw for lw in self.data.values() if lw]:
             self._highest_z = max(item.highest_z, self._highest_z)
 
-    def get_slot_definition(self, slot_name) -> "SlotDefV3":
+    def get_slot_definition(self, slot_name: DeckLocation) -> SlotDefV3:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/protocol_api/deck.py]:
```suggestion
    def get_slot_definition(self, slot_name: str) -> SlotDefV3:
```

File: api/src/opentrons/protocol_api/core/protocol_api/deck.py [hunk 9]
```diff
@@ -192,17 +198,17 @@ def get_slot_definition(self, slot_name) -> "SlotDefV3":
             )
         return slot_def
 
-    def get_slot_center(self, slot_name) -> types.Point:
+    def get_slot_center(self, slot_name: DeckLocation) -> Point:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/protocol_api/deck.py]:
```suggestion
    def get_slot_center(self, slot_name: str) -> Point:
```

File: api/src/opentrons/protocol_api/core/protocol_api/deck.py (hunk 6 above)
Comment (mcous) [reply api/src/opentrons/protocol_api/core/protocol_api/deck.py]: Buggy all along, change in type definition reflects what was actually happening in the implementation

File: api/src/opentrons/protocol_api/core/protocol_api/deck.py [hunk 10]
```diff
@@ -146,14 +152,14 @@ def is_edge_move_unsafe(self, mount: types.Mount, target: "Labware") -> bool:
...
-        if mount is types.Mount.RIGHT:
+        if mount is Mount.RIGHT:
             other_labware = self.left_of(slot)
         else:
             other_labware = self.right_of(slot)
 
         return isinstance(other_labware, ModuleGeometry)
 
-    def right_of(self, slot: str) -> Optional[DeckItem]:
+    def right_of(self, slot: DeckLocation) -> Optional[DeckItem]:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/protocol_api/deck.py]:
```suggestion
    def right_of(self, slot: str) -> Optional[DeckItem]:
```

File: api/src/opentrons/protocol_api/core/protocol_api/deck.py [hunk 11]
```diff
@@ -162,7 +168,7 @@ def right_of(self, slot: str) -> Optional[DeckItem]:
             idx = int(slot) + 1
             return self[str(idx)]
 
-    def left_of(self, slot: str) -> Optional[DeckItem]:
+    def left_of(self, slot: DeckLocation) -> Optional[DeckItem]:
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/protocol_api/deck.py]:
```suggestion
    def left_of(self, slot: str) -> Optional[DeckItem]:
```

**Review by mcous — COMMENTED** [review]
Got a handful of type annotation changes to fix / revert in the legacy Deck still

File: api/src/opentrons/protocol_api/core/protocol_api/deck.py (hunk 3 above)
Comment (mcous) [reply api/src/opentrons/protocol_api/core/protocol_api/deck.py]: I've tried to get this to work in the past and failed. We're hampered by the fact that in Python 3.7, `UserDict` is not generic at run time. I'd prefer a type ignore

**Review by SyntaxColoring — APPROVED** [review]
Thanks!

**Pull request merged**